pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,227 @@
+//
+// Note:
+//  since this controller needs to reconcile only master nodes and machine objects
+//  make sure nodeInformer and machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func NewClusterMemberRemovalController(
+	isMachineAPIFunctional func() (bool, error),
+	etcdClient etcdcli.EtcdClient,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	hasNetworkInformerSyncedFn func() bool,
+	networkLister configv1listers.NetworkLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &ClusterMemberRemovalController{
+		etcdClient:                 etcdClient,
+		isMachineAPIFunctionalFn:   isMachineAPIFunctional,
+		masterNodeLister:           masterNodeLister,
+		masterMachineLister:        masterMachineLister,
+		networkLister:              networkLister,
+		hasNetworkInformerSyncedFn: hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	// we are only interested in unhealthy members
+	// because theirs nodes are gone
+	unhealthyMembers, err := c.etcdClient.UnhealthyMembers(ctx)
+	if err != nil {
+		return err
+	}
+	if len(unhealthyMembers) == 0 {
+		return nil
+	}
+	for _, unhealthyMember := range unhealthyMembers {
+		memberURL, err := memberToURL(unhealthyMember)
+		if err != nil {
+			klog.Errorf("%v", err)
+			continue
+		}
+		memberLocator := fmt.Sprintf("%v/%v/%v", memberURL, unhealthyMember.Name, unhealthyMember.ID)
+
+		machine, err := c.getMachineForMember(unhealthyMember)
+		if err != nil {
+			klog.Errorf("unable to get a machine for the unhealthy member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if machine != nil {
+			klog.V(4).Infof("cannot remove the unhealthy member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
+			continue
+		}
+
+		node, err := c.getNodeForMember(unhealthyMember)
+		if err != nil {
+			klog.Errorf("unable to get a node for the unhealthy member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove the unhealthy member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
```

Comment:
Node is likely to be removed before the Machine, it's marginal, but we do delete the Node from the Machine controller before we remove the Machine finalizer. Probably doesn't matter.

It's not obvious reading this why a Machine being present, but the Node not being present is a valid scenario where possibly the etcd is still running/could come back to healthy, might be worth a comment somewhere to explain why. (The only one I can think of is if a user deletes the Node object, and etcd is broken, and then kubelet is restarted and the Kubelet recreates the Node - Is that something we'd even support, the Node UID changes, anything relying on that?)

Reply:
> It's not obvious reading this why a Machine being present, but the Node not being present is a valid scenario

it is not a valid scenario, this controller will remove a member only when both (the machine and the node) are not present